core/client/fs/src/main/java/alluxio/client/block/stream/BufferCachingGrpcDataReader.java

```diff
@@ -0,0 +1,253 @@
+    }
+    Preconditions.checkState(mPosToRead - mReadRequest.getOffset() <= mReadRequest.getLength());
+    return buffer;
+  }
+
+  /**
+   * Closes the {@link BufferCachingGrpcDataReader}.
+   */
+  public void close() throws IOException {
+    if (mClient.get().isShutdown()) {
+      return;
+    }
+    mCloser.close();
+  }
+
+  /**
+   * Increases the reference count and return the current count.
+   *
+   * @return the current count
+   */
+  public int ref() {
+    return mRefCount.incrementAndGet();
+  }
+
+  /**
+   * Decreases the reference count and return the current count.
+   *
+   * @return the current count
+   */
+  public int deRef() {
+    return mRefCount.decrementAndGet();
+  }
+
+  /**
+   * @return the current count
+   */
+  public int getRefCount() {
+    return mRefCount.get();
+  }
+
+  /**
+   * Factory class to create {@link BufferCachingGrpcDataReader}s.
+   */
+  public static class Factory {
+    private final FileSystemContext mContext;
+    private final WorkerNetAddress mAddress;
+    private final ReadRequest mReadRequest;
+
+    /**
+     * Creates an instance of {@link BufferCachingGrpcDataReader.Factory} for block reads.
+     *
+     * @param context the file system context
+     * @param address the worker address
+     * @param readRequest the read request
+     */
+    public Factory(FileSystemContext context, WorkerNetAddress address,
+        ReadRequest readRequest) {
+      mContext = context;
+      mAddress = address;
+      mReadRequest = readRequest;
+    }
+
+    /**
+     * @return a new {@link BufferCachingGrpcDataReader}
+     */
+    public BufferCachingGrpcDataReader create() throws IOException {
+      AlluxioConfiguration alluxioConf = mContext.getClusterConf();
+      int readerBufferSizeMessages = alluxioConf
+          .getInt(PropertyKey.USER_STREAMING_READER_BUFFER_SIZE_MESSAGES);
+      long dataTimeoutMs = alluxioConf.getMs(PropertyKey.USER_STREAMING_DATA_TIMEOUT);
+
+      CloseableResource<BlockWorkerClient> client = mContext.acquireBlockWorkerClient(mAddress);
+
+      GrpcBlockingStream<ReadRequest, ReadResponse> stream;
+      try {
+        String desc = "BufferCachingGrpcDataReader";
+        if (LOG.isDebugEnabled()) { // More detailed description when debug logging is enabled
+          desc = MoreObjects.toStringHelper(this)
+              .add("request", mReadRequest)
+              .add("address", mAddress)
+              .toString();
+        }
+        // Stream here cannot be GrpcDataMessagingBlockingStream
+        // DataBuffer.getReadOnlyByteBuffer is used to clone a copy in SharedDataReader.readChunk.
+        // getReadOnlyByteBuffer is not implemented in DataBuffer
+        // returned from GrpcDataMessagingBlockingStream.
+        stream = new GrpcBlockingStream<>(client.get()::readBlock, readerBufferSizeMessages,
+            desc);
+        stream.send(mReadRequest, dataTimeoutMs);
+      } catch (Exception e) {
+        client.close();
```

Comment:
do we need to close stream here too?

Reply:
Add the stream close here, 
I am little worry about when creating the stream, errors thrown, but i guess can still do the close() since it checks isOpen first.